main.go

```diff
@@ -14,191 +14,336 @@ import (
+					ConstLabels: prometheus.Labels{
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			cpuSysVal.Set(stat.Statistics.CpusSystemTimeSecs)
+
+			cpuUsrVal := prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Namespace: "mesos_task",
+					Name:      "cpu_user_seconds_total",
+					Help:      "Cumulative user CPU time in seconds.",
+					ConstLabels: prometheus.Labels{
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			cpuUsrVal.Set(stat.Statistics.CpusUserTimeSecs)
+
+			memLimitVal := prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Namespace: "mesos_task",
+					Name:      "memory_limit_bytes",
+					Help:      "Task memory limit in bytes.",
+					ConstLabels: prometheus.Labels{
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			memLimitVal.Set(float64(stat.Statistics.MemLimitBytes))
+
+			memRssVal := prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Namespace: "mesos_task",
+					Name:      "memory_rss_bytes",
+					Help:      "Task memory RSS usage in bytes.",
+					ConstLabels: prometheus.Labels{
+						"task":         stat.Source,
+						"mesos_slave":  host,
+						"framework_id": stat.FrameworkId,
+					},
+				},
+			)
+			memRssVal.Set(float64(stat.Statistics.MemRssBytes))
+
+			outCh <- cpuLimitVal
+			outCh <- cpuSysVal
+			outCh <- cpuUsrVal
+			outCh <- memLimitVal
+			outCh <- memRssVal
+		}
 	}
 }
 
-func processRequests(c *Config) {
-	for _, endpoint := range *c.Endpoints {
-		glog.Infof("Found %s\n", endpoint.Name)
+func (e *PeriodicExporter) lockSlaves(f func() []string) []string {
+	e.slaves.Lock()
+	defer e.slaves.Unlock()
+	return f()
+}
 
-		if len(endpoint.Interval) == 0 {
-			endpoint.Interval = c.Interval
-		}
+func (e *PeriodicExporter) rLockMetrics(f func()) {
+	e.RLock()
+	defer e.RUnlock()
+	f()
+}
 
-		if len(endpoint.TimeoutStr) != 0 {
-			t, err := time.ParseDuration(endpoint.TimeoutStr)
-			if err != nil {
-				glog.Fatalf("Invalid specification of time.Duration: %s\n", err)
-			}
-			endpoint.Timeout = t
-		} else {
-			endpoint.Timeout = 3 * time.Second
+func (e *PeriodicExporter) runEvery(f func(), interval time.Duration) {
+	f()
+	for {
+		select {
+		case <-time.After(interval):
+			f()
 		}
+	}
+}
 
-		go requestRunner(endpoint)
+func (e *PeriodicExporter) setMetrics(ch chan prometheus.Gauge) {
+	metrics := make([]prometheus.Gauge, 0)
+	for metric := range ch {
+		metrics = append(metrics, metric)
 	}
+
+	e.Lock()
+	e.metrics = metrics
+	e.Unlock()
 }
 
-func init() {
-	prometheus.MustRegister(cpuLimitVal)
-	prometheus.MustRegister(cpuSysVal)
-	prometheus.MustRegister(cpuUsrVal)
-	prometheus.MustRegister(memLimitVal)
-	prometheus.MustRegister(memRssVal)
+func (e *PeriodicExporter) scrapeSlaves() {
+	hostnames := e.lockSlaves(func() []string {
```

Comment:
This will create a new slice, but it will point to the same backing array as `e.slaves.hostnames`. If that is changed via the original slice reference, this is not safe... it seems like whenever you change the original slice, you replace it by a completely new one, so this shouldn't break in this specific case, but I'd still always make a copy of an entire slice if you want to work with it without a lock (except if you make it very explicit that the underlying array will never be modified).